src/chipper/keyboard2.cljs

```diff
@@ -0,0 +1,233 @@
+  #{:Space :ArrowDown :ArrowUp :ArrowLeft :ArrowRight :Tab :Backspace})
+
+(defn prevent-movement! [ev]
+  "Prevent movement keys (arrows, space, tab) from moving focus or scrolling."
+  (when (-movement-keys (keyword (.-code ev)))
+    (.preventDefault ev))
+  ev)
+
+(defn translate-keycode [ev]
+  "Takes keydown event and returns keycode as keyword; e.g., :KeyI, :Digit0.
+  If SHIFT is held, prepend Shift; e.g., :ShiftKeyI.
+  If CTRL is held, prepend Ctrl; e.g., :CtrlShiftKeyI."
+  (keyword (str (when (.-ctrlKey ev) "Ctrl")
+                (when (.-shiftKey ev) "Shift")
+                ;; .-code is nil if not a keypress
+                (.-code ev))))
+
+(defn set-position!
+  "Cursor position setter. Prefer using this over swap!-ing the state
+  atom directly"
+  [[line chan attr] state]
+  (swap! state assoc
+         :active-line line
+         :active-chan chan
+         :active-attr attr))
+
+(defn set-relative-position!
+  "This is still a mess, sort of"
+  [[dline dchan dattr] state]
+  ;; ever heard of select-keys?
+  (let [line (:active-line @state)
+        chan (:active-chan @state)
+        attr (:active-attr @state)
+        next-line (u/bounded-add (dec frame-length) line dline)
+        ;; TODO temporary, hopefully :)
+        ;; for now, we'll use cursor-attr-pos for these next two until the
+        ;; composer state representation gets fixed...
+        cursor-attr-pos (+ (* attr-count chan) attr)
+        ;; a plain mod of attr + dattr would cause the cursor to cycle in the first
+        ;; and last channel positions, so this is one way to solve that
+        next-attr- (u/bounded-add (dec total-attr-positions) cursor-attr-pos dattr)
+        next-attr (mod next-attr- attr-count)
+        next-chan (quot next-attr- attr-count)]
+    (set-position! [next-line next-chan next-attr] state)))
+
+;; it's just gotta map the keywords to the numbahz
+;; and that's just because all the bounds checking happens here; the ui just
+;; silently fails to render the cursor (and probably other things) if you give
+;; invalid indices
+(defn set-absolute-position!
+  "bit of potentially confusing naming, given the existence of set-position!"
+  [params state]
+  (js/alert "not implemented!"))
+
+(defn set-frame!
+  "Frame setter. Prefer using this over swap!-ing the state atom directly."
+  ;; I hate that I'm using the word setter and defining functions called set-*
+  [frame state]
+  (swap! state assoc
+         :active-frame frame))
+
+;; TODO set frame-edited
+(defn set-relative-frame!
+  "Frame setter. Prefer using this over swap!-ing the state atom directly."
+  ;; I hate that I'm using the word setter and defining functions called set-*
+  [dframe state]
+  (let [frame (:active-frame @state)]
+    (set-frame!
+      (u/bounded-add (dec frame-count) frame dframe)
+      state)))
+
+(defn set-attr!
+  "Attribute setter. Prefer using this over swap!-ing the state atom directly."
+  [params state])
+
+(defn set-attr-at-cursor!
```

Comment:
Should be possible to just tear this out of the original keyboard thing with minor modifications. Also, would be a good idea to add another helper for getting the cursor position, like https://github.com/cqsd/chipper/pull/24/files#diff-88e8e87d0e646860028f7e537449e7cbR158